File: sdk/prismer-cloud/python/prismer/aip.py
"""Prismer SDK — AIP Identity (Platform Integration)

Re-exports aip-sdk core. Install both packages:
    pip install prismer aip-sdk

Usage (standalone):
    from prismer.aip import AIPIdentity
    identity = AIPIdentity.create()

Usage (platform integration — planned v1.7.4):
    from prismer.aip import PrismerAIPAgent
    agent = PrismerAIPAgent.register(client, api_key)
"""

# Re-export from standalone AIP SDK
from aip import (  # type: ignore[import-untyped]
    AIPIdentity,
    public_key_to_did_key,
    did_key_to_public_key,
    validate_did_key,
    build_credential,
    verify_credential,
    build_presentation,
    verify_presentation,
    build_delegation,
    build_ephemeral_delegation,
    verify_delegation,
    verify_ephemeral_delegation,
)

__all__ = [
    "AIPIdentity",
    "PrismerAIPAgent",
    "public_key_to_did_key", "did_key_to_public_key", "validate_did_key",
    "build_credential", "verify_credential", "build_presentation", "verify_presentation",
    "build_delegation", "build_ephemeral_delegation", "verify_delegation", "verify_ephemeral_delegation",
]


class PrismerAIPAgent:
    """Wraps AIPIdentity with Prismer IM platform registration (v1.8.0 S7).

    Example::

        from prismer import PrismerClient
        from prismer.aip import PrismerAIPAgent

        client = PrismerClient(api_key="sk-prismer-...")
        agent = PrismerAIPAgent.register(client, "sk-prismer-...")
        # agent.did → 'did:key:z6Mk...'
    """

    def __init__(self, identity: "AIPIdentity"):
        self.identity = identity
        self._registered = False

    @classmethod
    def register(cls, client, api_key: str) -> "PrismerAIPAgent":
        """Create agent from API key and register with Prismer IM."""
        identity = AIPIdentity.from_api_key(api_key)
        agent = cls(identity)
        agent.ensure_registered(client)
        return agent

    @classmethod
    def from_private_key(cls, client, private_key_b64: str) -> "PrismerAIPAgent":
        """Create agent from Base64 private key and register."""
        import base64
        key_bytes = base64.b64decode(private_key_b64)
        identity = AIPIdentity.from_private_key(key_bytes)
        agent = cls(identity)
        agent.ensure_registered(client)
        return agent

    def ensure_registered(self, client) -> None:
        """Register identity key with IM server (idempotent)."""
        if self._registered:
            return
        try:
            client.im.identity.register_key(self.identity.public_key_base64)
            self._registered = True
        except Exception as e:
            if "already" in str(e).lower() or getattr(e, "status_code", 0) == 409:
                self._registered = True
            else:
                raise

    @property
    def did(self) -> str:
        return self.identity.did

    @property
    def public_key_base64(self) -> str:
        return self.identity.public_key_base64

File: sdk/prismer-cloud/typescript/src/aip.ts
/**
 * Prismer SDK — AIP Identity (Platform Integration)
 *
 * Re-exports @prismer/aip-sdk core and adds platform-specific convenience:
 * - All standalone AIP SDK types and classes are re-exported
 * - PrismerAIPAgent adds platform registration (auto-register DID on init)
 *
 * Usage (standalone — no Prismer platform needed):
 *   import { AIPIdentity } from '@prismer/sdk';
 *   const id = await AIPIdentity.create();
 *
 * Usage (platform integration — registers DID with Prismer IM):
 *   import { PrismerAIPAgent } from '@prismer/sdk';
 *   const agent = await PrismerAIPAgent.register(client, apiKey);
 */

// ─── Re-export everything from standalone AIP SDK ───────────
// Users of @prismer/sdk get full AIP SDK API without installing separately
export {
  AIPIdentity,
  type DIDDocument,
  type SignedPayload,
} from '@prismer/aip-sdk';

export {
  publicKeyToDIDKey,
  didKeyToPublicKey,
  validateDIDKey,
} from '@prismer/aip-sdk';

export {
  type VerifiableCredential,
  type VerifiablePresentation,
  buildCredential,
  buildPresentation,
  verifyCredential,
  verifyPresentation,
} from '@prismer/aip-sdk';

export {
  type VerifiableDelegation,
  type EphemeralDelegation,
  buildDelegation,
  buildEphemeralDelegation,
  verifyDelegation,
  verifyEphemeralDelegation,
} from '@prismer/aip-sdk';

// ─── Platform-specific integration (v1.8.0 S1) ────────────

import { AIPIdentity } from '@prismer/aip-sdk';

/**
 * PrismerAIPAgent — wraps AIPIdentity with platform registration.
 *
 * Auto-registers the DID with Prismer IM on creation, making the
 * agent's identity available for message signing and verification.
 *
 * @example
 * ```typescript
 * const client = new PrismerClient({ apiKey: 'sk-prismer-...' });
 * const agent = await PrismerAIPAgent.register(client, 'sk-prismer-...');
 * // agent.identity.did → 'did:key:z6Mk...'
 * // Messages sent via client.im will auto-sign if identity config is set
 * ```
 */
export class PrismerAIPAgent {
  readonly identity: AIPIdentity;
  private _registered = false;

  private constructor(identity: AIPIdentity) {
    this.identity = identity;
  }

  /** Create agent from API key + auto-register with Prismer IM */
  static async register(client: any, apiKey: string): Promise<PrismerAIPAgent> {
    const identity = await AIPIdentity.fromApiKey(apiKey);
    const agent = new PrismerAIPAgent(identity);
    await agent.ensureRegistered(client);
    return agent;
  }

  /** Create agent from existing private key + auto-register */
  static async fromPrivateKey(client: any, privateKeyBase64: string): Promise<PrismerAIPAgent> {
    const keyBytes = typeof Buffer !== 'undefined'
      ? new Uint8Array(Buffer.from(privateKeyBase64, 'base64'))
      : new Uint8Array(atob(privateKeyBase64).split('').map(c => c.charCodeAt(0)));
    const identity = await AIPIdentity.fromPrivateKey(keyBytes);
    const agent = new PrismerAIPAgent(identity);
    await agent.ensureRegistered(client);
    return agent;
  }

  /** Register identity key with the IM server (idempotent) */
  async ensureRegistered(client: any): Promise<void> {
    if (this._registered) return;
    try {
      await client.im.identity.registerKey(this.identity.publicKeyBase64);
      this._registered = true;
    } catch (err: any) {
      // Already registered is OK
      if (err?.message?.includes('already') || err?.status === 409) {
        this._registered = true;
      } else {
        throw err;
      }
    }
  }

  get did(): string { return this.identity.did; }
  get publicKeyBase64(): string { return this.identity.publicKeyBase64; }
}
